### DISCUSSION TITLE
Subclass of pydantic.BaseModel is not recognized as Body (json) parameter

### DISCUSSION THREAD

**Discussion by rigens** [fastapi/fastapi#12038] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import collections

import uvicorn
from fastapi import FastAPI
from pydantic import RootModel


class Strings(RootModel[list[str]], collections.abc.Sequence):
    def __getitem__(self, index):
        return self.root[index]

    def __len__(self):
        return len(self.root)

    def __iter__(self):
        yield from self.root


app = FastAPI()


@app.post('/post-strings')
async def post_strings(strings: Strings):
    return strings


uvicorn.run(
    app,  # type: ignore
    host='127.0.0.1',
    port=8088,
)
```

### Description

`Strings` is a subclass of `pydantic.BaseModel` and should be automatically recognized as `Body` parameter. But `FastAPI` treats it as `Form` parameter

```bash
curl -X POST http://127.0.0.1:8088/post-strings -d '["foo","bar","baz"]' -H "Content-Type: application/json"
...
< HTTP/1.1 422 Unprocessable Entity
{"detail":[{"type":"missing","loc":["body","strings"],"msg":"Field required","input":null}]}
```

The reason for this behavior is incorrect implementation of the `_compat.is_uploadfile_sequence_annotation` function.

Related PR: https://github.com/fastapi/fastapi/pull/10393

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.8.2

### Python Version

3.12.3

**Comment by ecly** [comment]
@rigens while I'm not certain if this is intended behavior, this seems to be a result of the `collections.abc.Sequence` subclassing, causing the `FieldInfo` to become an instance of `fastapi.params.File` instead of `fastapi.params.Body`. 

Bit of a partial solution, but I'd omit that sequence subclasses and if you need to for type checking, simply pass around the `root` object instead of the entire `RootModel` for the rest of your application.

**Comment by rigens** [comment]
@ecly, this is a BUG as described here https://github.com/fastapi/fastapi/pull/10393

> Bit of a partial solution, but I'd omit that sequence subclasses and if you need to for type checking, simply pass around the root object instead of the entire RootModel for the rest of your application

thanks, but we need this subclassing

**Comment by rigens** [comment]
Clearly this is a BUG. 
@tiangolo, please create an issue from this discussion.

**Comment by tiangolo** [comment, marked as answer]
Hello @rigens, you can achieve the same with:

```Python
from fastapi import FastAPI

app = FastAPI()


@app.post("/post-strings")
async def post_strings(strings: list[str]):
    return strings
```

As FastAPI allows you to use types used for Pydantic fields, there's no need to use `RootModel`, it's not officially supported, because you can achieve the same in better and simpler ways.

If you need to modify things further, you can also use additional Pydantic validators: https://docs.pydantic.dev/latest/concepts/validators/

**Comment by rigens** [comment]
> ...you can achieve the same with...

It's not the same... We need the custom pydantic model (not list[str]) for some other purposes dictated by our application logic (not only for FastAPI request validation) 

>  ...RootModel, it's not officially supported...

Really? `RootModel` is just a subclass of `BaseModel`, and should be treated the same way as any other pydantic model

@tiangolo, I have clearly pointed out the BUG that is causing the incorrect behavior, and instead of fixing it, you suggest that I use  some workarounds
